Makefile: Check that $BUILDSYS_ARCH is a supported architecture

zmrow opened the pull request:
**Issue number:**
Related to #1162 

**Description of changes:**
```
$BUILDSYS_ARCH is used all over the place in the Makefile, and if we don't
check that it's a valid supported architecture first, spurious directories get
created and commands further down the line will fail.  This change checks
$BUILDSYS_ARCH first before anything happens and fails.

    
It doesn't use any parameter expansion like the previous check did, meaning it
will be compatible with all versions of bash.
```

**Testing done:**
* `cargo make` commands fail if an invalid arch is supplied via `-e BUILDSYS_ARCH=`
```
...
[cargo-make] INFO - Running Task: setup
Unrecognized architecture 'bar'; please use 'x86_64 or 'aarch64'
...
```

* `cargo make` commands with valid architectures work as expected.


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

zmrow committed:
Makefile: Check that $BUILDSYS_ARCH is a supported architecture

$BUILDSYS_ARCH is used all over the place in the Makefile, and if we don't
check that it's a valid supported architecture first, directories get created
and commands further down the line will fail.  This change checks
$BUILDSYS_ARCH first before anything happens and fails.

It doesn't use any parameter expansion like the previous check did, meaning it
will be compatible with all versions of bash.

zmrow commented:
^ Fixed a typo

tjkirch commented on Makefile.toml:
```diff
@@ -98,6 +98,15 @@ PUBLISH_AMI_NAME_DEFAULT = "${BUILDSYS_NAME}-${BUILDSYS_VARIANT}-${BUILDSYS_ARCH
 '''
+# Ensure we use a supported architecture
+case "${BUILDSYS_ARCH}" in
+   x86_64|aarch64)
+   ;;
+   *)
+      echo "Unrecognized architecture '${BUILDSYS_ARCH}'; please use 'x86_64 or 'aarch64'"
+      exit 1 ;;
```
```suggestion
   x86_64|aarch64) ;;
   *)
      echo "Unrecognized architecture '${BUILDSYS_ARCH}'; please use 'x86_64 or 'aarch64'"
      exit 1
      ;;
```
nit: a more standard style.

zmrow commented:
^ style points

izaakschroeder commented:
Exciting, thank-you! ❤️

bcressey commented on Makefile.toml:
```diff
@@ -437,16 +446,6 @@ fi
 lz4 -df "${rootlz4}" "${root_image}"
 lz4 -df "${datalz4}" "${data_image}"
 
-# Canonicalize architecture identifier to the value recognized by EC2.
-case "${BUILDSYS_ARCH,,}" in
-   arm64|aarch64)
-      arch=arm64 ;;
-   x86_64|amd64)
-      arch=x86_64 ;;
-   *)
-      arch="${BUILDSYS_ARCH}" ;;
-esac
```
If we don't already, we need to handle at least the `aarch64` -> `arm64` canonicalization in pubsys, since it's going away here.

zmrow replied:
This is fixed by using the `parse_arch` function the SSM code already uses to automatically canonicalize the arch to the correct value.

zmrow replied:
Fixed using the `parse_arch` function that automatically canonicalizes the architecture to the proper value.

zmrow committed:
pubsys: Canonicalize architecture for ec2 when building AMIs

When building AMIs, we need to canonicalize architectures to something that EC2
understands. This change uses an existing function to canonicalize the
architecture when running `pubsys ami`.

zmrow commented:
Fixes @bcressey 's concern - when we wrote the SSM publishing code we added a function to canonicalize architecture but hadn't gone back to add it to the AMI code.
